[Close Index API] Add unique UUID to ClusterBlock

tlrx opened the pull request:
Note: this pull request will be merged in the close-index-api-refactoring branch

This pull request add a unique id to cluster blocks, so that they can be uniquely identified if needed. This is important for the Close Index API where multiple concurrent closing requests can be executed at the same time. By adding a UUID to cluster block, we can generate unique "index closed" blocks that can be verified on shards and then checked again from the cluster state before closing the index.

elasticmachine commented:
Pinging @elastic/es-distributed

ywelsch commented on server/src/main/java/org/elasticsearch/action/admin/indices/close/TransportVerifyShardBeforeCloseAction.java:
```diff
@@ -79,25 +82,25 @@ protected void acquireReplicaOperationPermit(final IndexShard replica,
     }
 
-    private void executeShardOperation(final IndexShard indexShard) {
+    private void executeShardOperation(final ShardRequest request,final IndexShard indexShard) {
```
space missing

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java:
```diff
@@ -404,6 +415,17 @@ public Builder removeIndexBlock(String index, ClusterBlock block) {
+                return this;
+            }
+            indices.get(index).removeIf(block -> block.id() == blockId);
+            if (indices.get(index).isEmpty()) {
```
perhaps avoid doing the lookup of `index` twice?

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java:
```diff
@@ -148,6 +147,18 @@ public boolean hasIndexBlock(String index, ClusterBlock block) {
         return indicesBlocks.containsKey(index) && indicesBlocks.get(index).contains(block);
     }
 
+    public boolean hasIndexBlock(String index, int blockId) {
```
perhaps rename this to `hasIndexBlockWithId`?

tlrx replied:
Should I also rename `hasGlobalBlock(int blockId)` to `hasGlobalBlockWithId(int blockId)` in a follow up PR against master ?

ywelsch replied:
yes, either here or in a follow-up.

tlrx replied:
I opened https://github.com/elastic/elasticsearch/pull/36941

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java:
```diff
@@ -404,6 +415,17 @@ public Builder removeIndexBlock(String index, ClusterBlock block) {
             return this;
         }
 
+        public Builder removeIndexBlock(String index, int blockId) {
```
same here, rename this to `removeIndexBlockWithId`

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -220,9 +223,7 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final ClusterSta
-            if (currentState.blocks().hasIndexBlock(index.getName(), INDEX_CLOSED_BLOCK) == false) {
-                blocks.addIndexBlock(index.getName(), INDEX_CLOSED_BLOCK);
-            }
+            blocks.addIndexBlock(index.getName(), closingBlock);
```
what if the index already has a close block? Do we keep on adding these blocks for every close request? Is there a risk here for the cluster state to become very large in case of many close requests?
Should we replace the current close block by a new one or perhaps limit the number of close blocks?

tlrx replied:
Hum. I think that we should always keep the last index closed block only and replace the current close block by a new one.

ywelsch replied:
I think we have different options here, with different tradeoffs.
- when starting a close on an index that is not yet fully closed, replace the current closed block by a new one. That means that the possibly ongoing close action cannot successfully complete and the new one will be the one to successfully complete. This is problematic because concurrent calls to close will now make them fail.
- add a new closed block next to the already existing one. Opening will erase all closed blocks. That allows concurrent closes to successfully complete, but possibly adds many blocks to cluster state.
- reuse the existing closed block if there is already one. This combines the advantages of both solutions above.

tlrx replied:
>     * reuse the existing closed block if there is already one. This combines the advantages of both solutions above.

It is indeed more elegant but if multiple close actions share the same unique block the actions become kind of dependant of each others, no? If the first response received is a failure (let's say, a shard is relocating and in INITIALIZING state) the index block should be removed, but we could have potential responses with the same block that arrive later and they won't be able to succeed.

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -349,28 +354,33 @@ private void sendVerifyShardBeforeCloseRequest(final IndexShardRoutingTable shar
                 final IndexMetaData indexMetaData = metadata.getSafe(index);
-                if (indexMetaData.getState() != IndexMetaData.State.CLOSE) {
-                    if (result.getValue().isAcknowledged()) {
+                assert currentState.blocks().hasIndexBlock(index.getName(), closingBlock);
```
how do we know that the close block has not been removed? An open command could have come in in the mean-time, removing the close block?

tlrx replied:
I changed the logic here to accommodate with keeping only the last index closed block.

> An open command could have come in in the mean-time, removing the close block?

Opening indices only removes blocks and changes state of indices that are not OPEN so I think we're good and block should not be removed by an open command.

tlrx commented:
@ywelsch I've updated the code. Can you have another look please?

tlrx commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -349,28 +357,37 @@ private void sendVerifyShardBeforeCloseRequest(final IndexShardRoutingTable shar
-                if (indexMetaData.getState() != IndexMetaData.State.CLOSE) {
-                    if (result.getValue().isAcknowledged()) {
-                        logger.debug("closing index {} succeed, removing index routing table", index);
+                if (acknowledged) {
```
I tried to make this logic easy to understand but we can maybe do better, any suggestion welcome.

I also think that if we go with always keeping a single block in cluster state we should compute the AcknowledgedResponse by diffing the current state and the updated state.

tlrx commented:
@ywelsch I updated the code according to our last discussion:

- when concurrent close actions are executed, a unique per index closing block is reused if it exists
- the closing block remains if the close action failed
- the unique closing block is replaced by the existing INDEX_CLOSED_BLOCK once close action succeeded
- open actions removes all closing blocks

I also added a bit more tests. Let me know what you think! Thanks

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -220,19 +224,35 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final ClusterSta
+
+            ClusterBlock indexBlock = null;
+            final Set<ClusterBlock> clusterBlocks = currentState.blocks().indices().get(index.getName());
+            if (clusterBlocks != null) {
```
maybe add a method `getIndexBlockWithId` to ClusterBlocks that returns a Nullable ClusterBlock?

tlrx replied:
Sure

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -220,19 +224,35 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final ClusterSta
+                indexBlock = INDEX_CLOSED_BLOCK;
             }
-            blockedIndices.add(index);
+            if (indexBlock == null) {
```
put this in an `else` to the `useDirectClose` and assert that indexBlock has a UUID

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -349,29 +371,40 @@ private void sendVerifyShardBeforeCloseRequest(final IndexShardRoutingTable shar
-                } else {
-                    logger.debug("index {} has been closed since it was blocked before closing, ignoring", index);
+                if (indexMetaData.getState() == IndexMetaData.State.CLOSE) {
+                    logger.debug("closing index {} succeed  but index is already closed", index);
```
s/succeed/succeeded/ and also remove extra space

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -349,29 +371,40 @@ private void sendVerifyShardBeforeCloseRequest(final IndexShardRoutingTable shar
                 }
+                final ClusterBlock closingBlock = blockedIndices.get(index);
+                if (currentState.blocks().hasIndexBlock(index.getName(), closingBlock) == false) {
+                    logger.debug("closing index {} succeed but block has been removed in the mean time", index);
```
succeeded

ywelsch replied:
"meantime" in one word

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -349,29 +371,40 @@ private void sendVerifyShardBeforeCloseRequest(final IndexShardRoutingTable shar
+                    continue;
+                }
+
+                logger.debug("closing index {} succeed", index);
```
succeeded

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -349,29 +371,40 @@ private void sendVerifyShardBeforeCloseRequest(final IndexShardRoutingTable shar
+                metadata.put(IndexMetaData.builder(indexMetaData).state(IndexMetaData.State.CLOSE));
+                routingTable.remove(index.getName());
+                closedIndices.add(index.getName());
+
```
remove extra newline

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -494,4 +536,16 @@ private static int getTotalShardCount(ClusterState state, Index index) {
+     * @return Generates a {@link ClusterBlock} that blocks read and write operations on soon-to-be-closed indices. The
+     * cluster block is generated with the id value equals to {@link #INDEX_CLOSED_BLOCK_ID} and a unique UUID.
+     */
+    public static ClusterBlock createIndexClosedBlock() {
```
Maybe call this `createIndexClosingBlock`? It's a block that's used for the time where the index is in the process of closing, not already closed

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -494,4 +536,16 @@ private static int getTotalShardCount(ClusterState state, Index index) {
+            "time and writes operations are blocked in the meantime. Execute an open index request to unblock the index and allow writes " +
+            "operation again. Execute a new close index request will try to close the index again.";
+        return new ClusterBlock(INDEX_CLOSED_BLOCK_ID, UUIDs.randomBase64UUID(), description , false, false, false,
+            RestStatus.FORBIDDEN, ClusterBlockLevel.READ_WRITE);
```
should we only block writes here instead of both reads and writes? The error message will need updating if we are to also block reads...

tlrx replied:
Yes, we should only block writes.

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -494,4 +536,16 @@ private static int getTotalShardCount(ClusterState state, Index index) {
+     * cluster block is generated with the id value equals to {@link #INDEX_CLOSED_BLOCK_ID} and a unique UUID.
+     */
+    public static ClusterBlock createIndexClosedBlock() {
+        final String description = "Index is blocked due to on-going index closing operation. Note that the closing process can take " +
```
this suggests that there is always an active operation trying to close the index, which we know isn't in case of failures. I wonder if we should say something along the lines of "index preparing to close. Reopen the index to allow reads and writes again or retry closing the index to fully close the index".

tlrx replied:
Ok

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -349,29 +371,40 @@ private void sendVerifyShardBeforeCloseRequest(final IndexShardRoutingTable shar
+                final ClusterBlock closingBlock = blockedIndices.get(index);
+                if (currentState.blocks().hasIndexBlock(index.getName(), closingBlock) == false) {
+                    logger.debug("closing index {} succeed but block has been removed in the mean time", index);
+                    continue;
```
why has it succeeded here? The index is not closed and the block has disappeared, so it must have been reopened in the meanwhile. This should be counted as a failure, not a success.

tlrx replied:
Well, the message in the log is confusing. What succeeded here is the verification of shards using the block that has now disappeared. I changed to change the log message.

> This should be counted as a failure, not a success.

I agree. I pushed a change that combines the results of the TransportVerifyShardBeforeCloseAction and the cluster state changes in order to determine if the current close action has indeed closed the index.

ywelsch commented on server/src/test/java/org/elasticsearch/indices/state/ReopenWhileClosingIT.java:
```diff
@@ -0,0 +1,167 @@
+        final String indexName = "test";
+        createIndexWithDocs(indexName);
+
+        ensureClusterSizeConsistency();
```
what's the point of calling this?

ywelsch replied:
we guarantee that when the test method executes that the cluster is already fully formed

ywelsch commented on server/src/test/java/org/elasticsearch/indices/state/ReopenWhileClosingIT.java:
```diff
@@ -0,0 +1,167 @@
+            createIndexWithDocs(indices.get(i));
+        }
+
+        ensureClusterSizeConsistency();
```
idem

ywelsch commented on server/src/test/java/org/elasticsearch/indices/state/ReopenWhileClosingIT.java:
```diff
@@ -0,0 +1,167 @@
+        assertAcked(client().admin().indices().prepareOpen(reopenedIndices.toArray(Strings.EMPTY_ARRAY)));
+
+        releaseBlock.close();
+        closeIndexResponse.get();
```
should this throw an exception / convey a failed closing?

tlrx replied:
The close action should not be acknowledged so we can check that.

When we'll improve the Close Index API response we can decide if we want to throw an exception or report an error message. For now I think we should not change the API and just not ack the response.

ywelsch replied:
ok, can we assert that acked == false here then?

tlrx replied:
It has been changed to `assertFalse(closeIndexResponse.get().isAcknowledged());`, I think we're good

tlrx commented:
@ywelsch Thanks a lot! I've updated the code again.

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -242,10 +264,10 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final Map<Index,
                 // Create a new index closed block
-                indexBlock = createIndexClosedBlock();
+                indexBlock = createIndexClosingBlock();
+                assert Strings.hasLength(indexBlock.uuid()) : "Closing block should have a UUID";
```
this assertion should also cover the case where `indexBlock` was not previously null, i.e., rewrite this as:

```
} else {
  if (indexBlock == null) {
    indexBlock = createIndexClosingBlock();
  }
  assert Strings.hasLength(indexBlock.uuid()) : "Closing block should have a UUID";
}
```

tlrx replied:
Sure

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -142,11 +147,32 @@ public void clusterStateProcessed(final String source, final ClusterState oldSta
+                                                    if (result.getValue().isAcknowledged()) {
+                                                        if (updatedMetaData.getState() == IndexMetaData.State.CLOSE) {
+                                                            IndexMetaData previousMetaData = currentState.metaData().index(result.getKey());
+                                                            if (previousMetaData != null) {
```
how could this ever be null?

tlrx replied:
I can't, I tend to be too defensive on NPEs.

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java:
```diff
@@ -142,11 +147,32 @@ public void clusterStateProcessed(final String source, final ClusterState oldSta
+                                                        if (updatedMetaData.getState() == IndexMetaData.State.CLOSE) {
+                                                            IndexMetaData previousMetaData = currentState.metaData().index(result.getKey());
+                                                            if (previousMetaData != null) {
+                                                                acknowledged = (previousMetaData.getState() == IndexMetaData.State.OPEN);
```
I'm not sure I follow this logic here. Wouldn't it be much simpler just to say. If the index is closed at this point, all is good. If not, something must have gone wrong, thereby acked == false.

This amounts to the following simple logic:

```
for (Map.Entry<Index, AcknowledgedResponse> result : results.entrySet()) {
    IndexMetaData updatedMetaData = updatedState.metaData().index(result.getKey());
    if (updatedMetaData != null && updatedMetaData.getState() != IndexMetaData.State.CLOSE) {
        acknowledged = false;
        break;
    }
}
```

tlrx replied:
Agreed, let's do that.

tlrx commented:
Thanks @ywelsch  for the review. I fixed conflicts and updated the code. Let me know if you have more comments or concerns. Thanks

tlrx commented:
I pushed a fix for a test bug in `ReopenWhileClosingIT` (the test tried to execute an open index request by passing an empty list of indices). Other tests failures are caused by the target feature branch not being up to date with master.

tlrx commented:
Thanks @ywelsch, and sorry for the  numerous comments you had to make.
